### ISSUE TITLE
X stage becomes unreferenced after software start and can not be moved

### ISSUE THREAD

**Issue by raacampbell** [mesoSPIM/mesoSPIM-control#44]
I am on the latest version of the master branch (88de162150c4c549e77dc34e72a7af5b4e9263d7) and v5 stages. I start PI MikroMove and reference all stages. All servos are on and I confirm they move. I close MikroMove and then start mesoSPIM-control. The X stage no longer moves. In the console I see the error message : 

```
Unallowable move attempted on unreferenced axis, or attempted to move with servo off
```

It references line 145 of `mesoSPIM_Serial` and 387 of `mesoSPIM_Stages`. 

When I close the control software and re-reopen MikroMove I see that the X stage has become un-referenced. 

<img width="304" alt="PI_MM" src="https://user-images.githubusercontent.com/5427811/87036786-3fc78a00-c1e3-11ea-90b6-08600df1b8ce.PNG">

My x/min/max values are broad enough:
```
    'x_max' : 30000,
    'x_min' : 8500,
```

and I have:
```
pi_parameters = {'controllername' : 'C-884',
                 'stages' : ('L-509.20DG10','L-509.40DG10','L-509.20DG10','M-060.DG','M-406.4PD','NOSTAGE'),
                 'refmode' : ('FRF',),
                 'serialnum' : ('118045956'),
                 }
```

**Comment by raacampbell** [comment]
I have gone on to confirm the following:

1. Referencing in MikroMove then quitting and restarting MikroMove: stage is still referenced.
2. I added a check that all stages are referenced as early as possible in the `mesoSPIM_Stages` constructor (Pull Request: https://github.com/mesoSPIM/mesoSPIM-control/pull/46). The stage reports as being unreferenced even at this point. 
3. I can get the stage working by adding `self.pidevice.FRF(1)` to the constructor of `mesoSPIM_Stages`, but I don't know why this needed or if it's a good idea.

**Comment by ffvoigt** [comment]
Thanks for looking into this! Currently, I don't have access to a V5 setup so I can't duplicate this right now. However, I have the suspicion that the root cause is with PIPython. The referencing of the focus stage upon startup was caused by the focus stage loosing reference when starting early versions of the mesoSPIM-software - as noted in the source code, this is a hack. It was always the stage attached to the highest port that caused it - could it be that in your PIPython version, it is the lowest port instead? Ideally, we would not need re-referencing at all. 
In addition, referencing any sample stage is dangerous as there might be a collision risk (especially when restarting after a software crash) - whereas referencing the focus stage is usually ok.

**Comment by raacampbell** [comment]
Yes, that could be it. The focus stage is indeed referenced on startup (see pull request). I'll check later which version of PIPython I have. Can you let me know what you are running when you have a chance? Maybe we should force a specific version in the pip install line we specify. It looks like I can't swap the stages because `mesoSPIM_Stages` expects them to be connected to specific ports. 

I feel fairly safe with the reference motion, however, since the reference point is in the middle of the range. Have you seen dangerous behavior (e.g. moving to one extreme of the motion range in order to reference)?

**Comment by ffvoigt** [comment]
Sadly, PIPython is not on pip, so forcing a version is a little difficult when you have to use the version of the included installation DVD. 
The referencing behavior can be specified (I think it is the refmode setting) - but it also depends on the stage - the L-509 has reference switches at the limits and in the center of the travel range, but other PI stages don't (for example, the M-112K033 of the V4 has them at the end). So, it is hard to come up with a solution for all stage configurations - and by now, there is quite a variety of mesoSPIM stage variants.
I will check for my PIPython version and get back to you.

**Comment by raacampbell** [comment]
I see... What I didn't say is that I additionally upgraded the PC to Win 10 and so re-installed everything. I re-installed the PI software and also did the update, which seemed substantial. It would well be that other users run into the same problem going forward. I have a couple of suggestions:

1. Turn the `stages` parameter into a dictionary. Then we are free to change how the stages are connected. In addition, this makes the config file clearer and the control code more robust. 
2. Check on startup which stages are not referenced. If stages other than the focus stage need referencing, we can bring up a confirmation box. Perhaps if we know other stages that can be referenced safely then we can skip the dialog box for those too. 

I think both the above are not a bad idea in any case. What do you reckon? I don't mind doing this. For (1) is it only code in `mesoSPIM_Stages.py` which will need to modified?

**Comment by nvladimus** [comment]
Is this issue still active?
